### DISCUSSION TITLE
How to uuse Slonik in Fastify/Express?

### DISCUSSION THREAD

**Discussion by austinzentzkippdc** [gajus/slonik#390] [Q&A] [answered]
I'm really interested in Slonik as a database querying library, but I'm having a hard time understanding exactly how it would fit into a long-running back-end process like Express or Fastify. I'm hoping others who have figured out the right usage could share their insight.

From what I can figure out, the best approach is to establish a connection pool when the Express/Fastify server is created, and then to ensure that the pool is accessible subsequently when the server is handling requests. Then, when individual endpoints are called by users, they should establish a specific connection from the already instantiated pool, make the query, and then do whatever they want with the results.

Here's an example of what I _think_ this looks like in Fastify:

`slonik-plugin.ts`:
```ts
import { FastifyPluginAsync } from 'fastify';
import fp from 'fastify-plugin';
import { createPool, DatabasePool } from 'slonik';

declare module 'fastify' {
  interface FastifyInstance {
    pool: DatabasePool;
  }
}

export const slonikPlugin: FastifyPluginAsync = fp(async (server) => {
  const pool = await createPool(process.env.DATABASE_URL || '');
  server.decorate('pool', pool);
});
```

`app.ts`:
```ts
... [Build your server config]
fastify.register(slonikPlugin);
```

`route.ts`:
```ts
  /** Route:   GET /users
   *  Summary: Get list of users
   *  Details:
   */
  fastify.route<{
    Querystring: inputs.UserListOptions;
  }>({
    method: 'GET',
    url: '/',
    schema: {
      querystring: inputs.userListOptions,
      response: {
        200: outputs.userList,
      },
    },
    handler: async (request) => {
      const users = await fastify.pool.connect(async (connection) => {
        return connection.many(sql`SELECT * from "User" LIMIT 10`);
      });
      return { users }
  }
});
```

Is this more or less the recommended approach, or am I missing a more obvious way to get the initial configuration up and running? Are there simple abstractions I could take advantage of so that each route handler doesn't have to repeat the connection functionality?

**Comment by austinzentzkippdc** [comment]
After noodling on this a bit more, I think I prefer a slightly different setup. This uses a class to instantiate a Slonik Connection singleton, which is a thin wrapper on existing Slonik methods. Here's an example:

`connect.ts`
```ts
import {
  createPool,
  DatabasePool,
  TaggedTemplateLiteralInvocation,
} from 'slonik';

class Connection {
  pool?: DatabasePool;

  async getPool() {
    if (!this.pool) {
      this.pool = await createPool(process.env.DATABASE_URL || '');
      if (!this.pool) {
        throw new Error('Could not establish database connection pool');
      }
    }
    return this.pool;
  }

  async maybeOne(sql: TaggedTemplateLiteralInvocation) {
    const pool = await this.getPool();
    return pool.connect(async (connection) => {
      return connection.maybeOne(sql);
    });
  }

  async many(sql: TaggedTemplateLiteralInvocation) {
    const pool = await this.getPool();
    return pool.connect(async (connection) => {
      return connection.many(sql);
    });
  }

  // And so forth for other slonik connection routes used in the app
}

export const connect = new Connection();
```

This allows for a much more readable set of code over in my route handlers:
`user-routes.ts`
```ts
import { connect } from '../db/query.ts';

fastify.route<{
    Querystring: inputs.UserListOptions;
  }>({
    method: 'GET',
    url: '/',
    schema: {
      querystring: inputs.userListOptions,
      response: {
        200: outputs.userList,
      },
    },
    handler: async (request) => {
      const { query } = request;
      const testUsers = await connect.many(sql`SELECT * from "User" LIMIT 10`);
      return testUsers;
    }
});
```

Or I could abstract related queries into a queries page:
`user-queries.ts`
```ts
import { connect } from '../db/query.ts';

export const getUsers = () => {
  return connect.many(sql`SELECT * from "User" LIMIT 10`);
};
```

`user-routes.ts`
```ts
import { getUsers } from './user-queries';

fastify.route<{
    Querystring: inputs.UserListOptions;
  }>({
    method: 'GET',
    url: '/',
    schema: {
      querystring: inputs.userListOptions,
      response: {
        200: outputs.userList,
      },
    },
    handler: async (request) => {
      const { query } = request;
      const testUsers = await getUsers();
      return testUsers;
  }
});
```

I think this approach has two advantages over my first post:
1. I can reuse the user-queries outside of the API server context more easily, if I want to perform a similar query in a standalone job, worker, etc.
2. The route handler syntax is a lot easier to parse as a developer who may not know the specifics of Slonik connection handling.

Just sharing my approach here both for the benefit of anyone with a similar question, and because I'd appreciate feedback from any current Slonik users who may be using this or another approach to use Slonik within an API server.

**Comment by AndrewJo** [comment]
I don't know about Fastify but I've created a middleware for Express.js for wrapping request handlers in a transaction.

With [express-slonik](https://github.com/AndrewJo/express-slonik) you can ensure your entire request is wrapped in a single database transaction which is really useful when you want to ensure all of the handlers and middleware are properly isolated for a given request context:

```typescript
import express, { json } from "express";
import createMiddleware, { IsolationLevels } from "express-slonik";

// middleware/currentUser.ts
// Middleware that gets the current user from database from request session data.
const currentUser = function () {
  return async function (req, res, next) {
    try {
      req.currentUser = await req.transaction.one(
        sql`SELECT * FROM users WHERE users.id = ${req.session.userId}`
      );
      next()
    } catch (error) {
      next(error);
    }
  }
};

// middleware/checkPermission.ts
// Middleware to check whether current user has permission to perform this action
const checkPermission = function (permissionName: string) {
  return async function (req, res, next) {
    try {
      const permissions = await req.transaction.manyFirst(sql`
        SELECT permissions.name
        FROM roles
        INNER JOIN permissions ON permissions.role_id = roles.id
        INNER JOIN user_roles ON user_roles.role_id = roles.id
        WHERE user_roles.user_id = ${req.session.userId}
      `);
      )
    } catch (error) {
      next(error);
    }
  };
};

// app.ts
(async function () {
  const pool = await createPool(process.env.DATABASE_URL);
  const app = express();
  const transaction = createMiddleware(pool);

  app
    .use(json())
    .patch(
      "/post/:id",

      // Start a transaction context (can also specify isolation levels if needed)
      transaction.begin(),

      // Since currentUser and checkPermissions are wrapped in the transaction, we know that
      // nothing can modify the user's permissions outside of current request context.
      currentUser(),
      checkPermission("update:post"),

      // By the time we get to the request handler, we have sufficient guarantees that the user in
      // the current request context has the permission to update the post and that no one else
      // is updating this post at the same time.
      async (req, res, next) => {
        try {
          await req.transaction.query(sql`
            UPDATE posts SET 
              posts.updated_by = ${req.session.userId},
              posts.updated_at = now(),
              -- Don't do this at production kids, we want the content sanitized!
              posts.content = ${req.body.content}
            WHERE posts.id = ${req.params.id}`
          );

          const updatedPost = await req.transaction.one(
            sql`SELECT * FROM posts WHERE posts.id = ${req.params.id}`
          );

          // Respond with updated post JSON
          res.json(updatedPost);
        } catch (error) {
          next(error);
        }
      },

      // By default, express-slonik autocommits the transaction when a response is sent to the client
      // by Express.js or autorollbacks whenever an error is detected in the request handler chain. 
      // You can also be explicit or have a finer control on when this happens by uncommenting the 
      // middleware below:
      // transaction.end()

      // Middleware inserted after transaction.end will not have access to req.transaction context and
      // will need to begin a new transaction context!
})();
```

Checkout the [README](https://github.com/AndrewJo/express-slonik#readme) for more documentation and examples. As always, I would like to know what you guys think about this pattern for Express.js apps.

**Reply by gajus** [reply]
This is not a good idea.

I didn't read the entire codebase, but just going by what has been said:

You want your transactions to be as short as possible. Anything over 1 second is a no-go in my book. Otherwise, you are risking a whole cascade of performance problems, with no easy way to refactor, given the assumption made about the request isolation level.

The general pattern for using Slonik with Fastify (or any other client with any other server) should be:

* start a server and create a connection pool
* pass connection pool to request context
* take a connection from the pool when needed/use transaction when needed

That simple

**Reply by AndrewJo** [reply]
@gajus I don't think a blanket statement about whether transactions for a request context being a bad idea is helpful. Yes, if all you're after is to squeeze the every last drop of performance out of your database, then long lived transactions are indeed an anti-pattern. But there's a reason banks will happily trade away fastest query performance for strong consistency. 

Likewise at my company, ensuring atomicity for a single request context in a middleware chain was deemed more important than fastest possible DB query performance where there are potentially many writes to a table row vs an occasional read that can cause non-repeatable reads in the middleware stack. In our particular use cases, request lifecycle at P95 is no longer than several hundred milliseconds, not over a second. Our customers don't care their API call lasted extra 50ms due to the transaction but does care a lot when the data they get back is inconsistent.

Premature optimization is the root of all evil so I'm sure competent software engineers are capable of profiling their code and make the necessary tradeoffs as necessary. That simple. :smile:

**Reply by flevi29** [reply]
@gajus So for instance if I have an endpoint that invokes and returns function `x` which takes a connection, and this endpoint is frequently called upon, the best practice is still to pass a new connection to every one of these endpoint calls from a connection pool? By extension, is it a bad practice to keep a connection open for the lifetime of the application itself? And even if I would keep a connection open throughout the applications life, would the performance gain be negligible, or non-existent?

**Reply by gajus** [reply]
@AndrewJo If you have a public facing API and you are doing request-level transactions, then you better hope that no one with bad intentions knows about it. Fair regarding "blanket statement", but given that the absolute vast majority of users should probably never do this, I'd rather use tone that directs reader to be cautious.

If you are doing this, at least make sure that you have very strict timeouts.

**Reply by gajus** [reply]
@flevi29 Andrew's comment was about transaction. How you handle connection pooling is a different topic and will depend on the payload type. Personally, I would steer away from a pattern that issues a connection per request and instead assign a connection when the application is being invoked. These connections are already locally pooled (with default settings), so you are not compromising on performance.

**Reply by flevi29** [reply]
@gajus I am not sure I entirely follow. Is the following code the right approach?
```typescript
// (pool is only ever created once)
const pool = await createPool(...);

// I use Nestjs so I never write things like this, in case I make a mistake
app.get("some-endpoint", async (req, res, ...) => {
  // ...
  // x takes a `DatabaseConnection`, queries any of something and returns it
  res.json(await x(pool));
  // or maybe like so if `x` would use connection multiple times
  res.json(await pool.connect((connection) => x(connection)));
  // ...
});
```

EDIT: I mean it must be, re-reading what you wrote in the first reply, it exactly means this, or so I interpret it. I guess I was just confused about the whole `pool.connect` thing, thought I should somehow wrap the whole thing in it, or maybe extract a `connection` from `pool.connect`, but thinking more about it the types of connections coming from `pool.connect` aren't meant to be long lived like that I suppose, even if it's more efficient to use them multiple times when possible.

**Reply by gajus** [reply]
`res.json(await x(pool))` is correct.

**Comment by gajus** [comment, marked as answer]
Two articles on the subject:

* https://dev.to/gajus/integrating-slonik-with-expressjs-33kn
* https://dev.to/gajus/integrating-fastify-with-slonik-5bke
